deepspeed/utils/tensor_fragment.py

```diff
@@ -207,6 +216,26 @@ def safe_get_full_grad(param):
     return None
 
 
+def safe_set_full_grad(param, value):
+    """Update the partitioned gradient of a low-precision (e.g., fp16) parameter.
+
+        Args:
+            param (``torch.nn.Parameter``): A model parameter
+            value (``torch.Tensor``): New value
+    """
+    if param.grad is not None:
+        param.grad.copy_(value)
+        return
+
+    # ZeRO stage 3 param
```

Comment:
Should we clarify that only one case is fired from `if hasattr(param, 'ds_id')` and `if hasattr(param, '_hp_mapping')` by if-elif? Will this silently be nop when Z0 or caught by `if param.grad is not None`?

Reply:
Yes, you are correct, and I will improve this for clarity.  

Below is my current thinking - glad to get your thoughts.  
```python
  if param.grad is not None: 
  # Z0 and Z1-default
  elif hasattr(param, 'ds_id'):
 # Z3
  elif hasattr(param, '_hp_mapping'):
 # Z2, Z1+hooks (e.g., fp32 grad accum)
 else:
  Noop or maybe assert? 
```